src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Cart\Context;
+
+use Sylius\Component\Cart\Context\CartContextInterface;
+use Sylius\Component\Cart\Context\CartNotFoundException;
+use Sylius\Component\Cart\Model\CartInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Core\Context\ShopperContextInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+final class ShopBasedCartContext implements CartContextInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $cartFactory;
+
+    /**
+     * @var ShopperContextInterface
+     */
+    private $shopperContext;
+
+    /**
+     * @var CartInterface
+     */
+    private $cart;
+
+    /**
+     * @param FactoryInterface $cartFactory
+     * @param ShopperContextInterface $shopperContext
+     */
+    public function __construct(FactoryInterface $cartFactory, ShopperContextInterface $shopperContext)
+    {
+        $this->cartFactory = $cartFactory;
+        $this->shopperContext = $shopperContext;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCart()
+    {
+        if (null !== $this->cart) {
+            return $this->cart;
```

Comment:
@pamil This is not for caching. You want a single "new" cart per request, right? Because you call ``getCart()`` multiple times when adding for example and then saving. So it is not really for perf. reasons, but for the way that cart context is supposed to work.